### DISCUSSION TITLE
Issue Setting `JsonNullable` Field with Instancio Generator

### DISCUSSION THREAD

**Discussion by RuslanNyko** [instancio/instancio#1209] [Q&A]
Hello,
I’m encountering an issue while using Instancio to generate instances of a class that contains a `JsonNullable` field. Below is my setup:

**Pojo Class:**

```java
public class Pojo {
    private JsonNullable<String> string;

    public JsonNullable<String> getString() {
        return string;
    }

    public void setString(JsonNullable<String> string) {
        this.string = string;
    }
}
```
**Instancio Generator Code:**

```java
Pojo pojo = Instancio.of(Pojo.class)
    .generate(Select.field(Pojo::getString), g -> g.oneOf("S", "P", "O"))
    .create();
```

**Exception Thrown:**

```
Can not set org.openapitools.jackson.nullable.JsonNullable field com.dto.Pojo.string to java.lang.String
    at java.base/jdk.internal.reflect.FieldAccessorImpl.throwSetIllegalArgumentException(FieldAccessorImpl.java:228)
    at java.base/jdk.internal.reflect.FieldAccessorImpl.throwSetIllegalArgumentException(FieldAccessorImpl.java:232)
    at java.base/jdk.internal.reflect.MethodHandleObjectFieldAccessorImpl.set(MethodHandleObjectFieldAccessorImpl.java:115)
```
    
**Question:**

How can I correctly use Instancio to generate and set values for a `JsonNullable<String>` field? Is it possible to utilize `InstancioServiceProvider` for this purpose, or is there another recommended approach?

Any guidance or examples on handling `JsonNullable` with `Instancio` would be greatly appreciated!

**Comment by armandino** [comment]
Hi @RuslanNyko,

You can achieve this using a `Scope`:

```java
Scope jsonNullableScope = Select.scope(Pojo::getString);

Pojo pojo = Instancio.of(Pojo.class)
    .generate(Select.allStrings().within(jsonNullableScope), g -> g.oneOf("S", "P", "O"))
    .create();
```

You could also use `InstancioServiceProvider` if you don't want to specify these values manually each time you create your POJO.

Please see these links for more details:

- https://www.instancio.org/user-guide/#selector-scopes
- https://github.com/orgs/instancio/discussions/593

**Reply by RuslanNyko** [reply]
Thanks for the answer, I need a general solution that supports many types. Is it possible to put this in a separate class, for example, which implements `InstancioServiceProvider`?

**Reply by armandino** [reply]
Yes, it should be possible to add a generator for `JsonNullable` in a `InstancioServiceProvider` implementation.

I'm not sure what you mean by a general solution since in the example, the expected values are strings "S", "P", "O". Can you provide additional examples?

**Reply by RuslanNyko** [reply]
I have two classes, `Pojo` and `Pojo2`, where some fields are of type `JsonNullable`. I want to know if it is possible to check if a generator has already been defined for a particular field and reuse it in an instance generation, such as with the `Instancio` library.

Here are the two classes:

```java
class Pojo {
  private JsonNullable<String> stringJsonNullable;
  private JsonNullable<Integer> integerJsonNullable;
  private JsonNullable<Long> longJsonNullable;
  private JsonNullable<BigDecimal> bigDecimalJsonNullable;
  private JsonNullable<List<String>> listJsonNullable;
}
```

```java
class Pojo2 {
  private JsonNullable<String> stringJsonNullable;
  private JsonNullable<List<Pojo>> listJsonNullable;
}
```
I am using the following code to generate instances of `Pojo2`:

```java
private Pojo2 generatePojo2(){
    return Instancio.of(Pojo2.class)
        .generate(Select.field("stringJsonNullable"), g -> g.text().pattern("#d#d#d"))
        .create();
}
```
In this example, a `String` will be generated for the stringJsonNullable field. However, since the field is of type `JsonNullable<String>,` I want to ensure that the generated value is wrapped in `JsonNullable`—i.e., the generated string should be `JsonNullable.of("value")`, not just a plain `string`.

**The Question:**

- Is it possible to customize the generator to automatically wrap the generated value in JsonNullable?

- Can I check if a generator has already been applied to a specific field (e.g., stringJsonNullable) and reuse it for fields of type JsonNullable?

**Reply by armandino** [reply]
> * Is it possible to customize the generator to automatically wrap the generated value in JsonNullable?

There's no feature that allows wrapping generated values automatically. Instead you will need to target strings within `JsonNullable` using selectors and scopes using `Select.allStrings().within(scope(Pojo::getStringJsonNullable))`.

Alternatively, you can define a `JsonNullable` generator or `Supplier`. You can find an example in this [post](https://github.com/orgs/instancio/discussions/593).

> * Can I check if a generator has already been applied to a specific field (e.g., stringJsonNullable) and reuse it for fields of type JsonNullable?

It's not possible to check if a generator has been defined for a specific field. If you want a `JsonNullable` generator to be applied automatically, then you could use the `InstancioServiceProvider`. You can still override specific fields as needed.

**Reply by RuslanNyko** [reply]
Thanks for the answer, I will try to use the scope
